### ISSUE TITLE
RayNM / rtcIntersectionNM wrapper for extended source

### ISSUE THREAD

**Issue by steo85it** [sampotter/python-embree#12]
Hi! I'd like to use the library to trace illumination of a terrain from an **extended** source (i.e., the Sun as seen from Mercury), similarly to what done in [the python-flux library](https://github.com/sampotter/python-flux/blob/master/flux/shape.py) .
I guess I could use a loop of `intersect1M` calls with different source-points, but it would probably be far less efficient than using a `RayNM` call ... any suggestion about how to implement this by tweaking the current wrappers (or informed recommendation if it is/isn't worth doing it)? Thanks!

**Comment by sampotter** [comment]
Embree has a bunch of different ways of doing ray intersections (`intersect1`, `intersect1M`, `intersectNM`, `intersectMp` off the top of my head).

It's not clear to me what the performance trade-offs of each of these are. For instance, it appears that several of these support the "ray stream" functionality for tracing coherent bundles of rays.

I think you should try setting up one `intersect1M` call for all of the rays you want to trace. That is, the name `intersect1M` suggests that `ray.org` should be set the same for all rays, but this isn't the case. You can set `ray.org` and `ray.dir` however you like for each ray.

The function `TrimeshShapeModel.check_vis_1_to_N` does indeed traces rays from one face, but note that `ray.org` is actually different for each ray (the start of the ray is perturbed along the ray away from the triangle centroid, hence `ray.org` is different for every ray).

It would be simple to create a new function `TrimeshShapeModel.check_vis_M_to_N` which just batches together a bunch of rays for different `check_vis_1_to_N` calls into a single `intersect1M call.

---

tl;dr for people who are not users of python-flux: the name `intersect1M` does not mean that you are tracing one ray from 1 to point to N other points.

**Comment by sampotter** [comment]
Wrapping `RayNM` and `rtcIntersectNM` is still potentially of interesting but this is a low priority for me. If someone needs this, please ping me here.

**Comment by steo85it** [comment]
Sounds like a good idea. After all, it's already what I'm doing for the "extended Sun" (ray tracing between each point of the source - or each position of the source at different epochs - and all facets within the same scene/intersectcontext call). Thanks!
